File: go/dnse/api_market.go
package dnse

import (
	"fmt"
	"strconv"
)

// GetSecurityDefinition retrieves securities definition updates.
func (c *Client) GetSecurityDefinition(symbol, boardID string, dryRun bool) (int, []byte, error) {
	query := make(map[string]string)
	if boardID != "" {
		query["boardId"] = boardID
	}
	return c.Request("GET", fmt.Sprintf("/price/%s/secdef", symbol), query, nil, nil, dryRun)
}

// GetOHLC retrieves OHLC data for a specific instrument over a given time range.
func (c *Client) GetOHLC(barType string, queryParams map[string]string, dryRun bool) (int, []byte, error) {
	query := make(map[string]string)
	for k, v := range queryParams {
		query[k] = v
	}
	query["type"] = barType
	return c.Request("GET", "/price/ohlc", query, nil, nil, dryRun)
}

// GetTrades retrieves historical trade data for a specific instrument.
func (c *Client) GetTrades(symbol, boardID, fromDate, toDate string, limit *int, order, nextPageToken string, dryRun bool) (int, []byte, error) {
	query := make(map[string]string)
	if boardID != "" {
		query["boardId"] = boardID
	}
	if fromDate != "" {
		query["from"] = fromDate
	}
	if toDate != "" {
		query["to"] = toDate
	}
	if limit != nil {
		query["limit"] = strconv.Itoa(*limit)
	}
	if order != "" {
		query["order"] = order
	}
	if nextPageToken != "" {
		query["nextPageToken"] = nextPageToken
	}
	return c.Request("GET", fmt.Sprintf("/price/%s/trades", symbol), query, nil, nil, dryRun)
}

// GetInstruments retrieves the list of available trading instruments and their metadata.
func (c *Client) GetInstruments(symbol, marketID, securityGroupID, indexName string, limit, page *int, dryRun bool) (int, []byte, error) {
	query := make(map[string]string)
	if symbol != "" {
		query["symbol"] = symbol
	}
	if marketID != "" {
		query["marketId"] = marketID
	}
	if securityGroupID != "" {
		query["securityGroupId"] = securityGroupID
	}
	if indexName != "" {
		query["indexName"] = indexName
	}
	if limit != nil {
		query["limit"] = strconv.Itoa(*limit)
	}
	if page != nil {
		query["page"] = strconv.Itoa(*page)
	}
	return c.Request("GET", "/instruments", query, nil, nil, dryRun)
}

// GetLatestTrade retrieves the most recent trade for a specific instrument.
func (c *Client) GetLatestTrade(symbol, boardID string, dryRun bool) (int, []byte, error) {
	query := make(map[string]string)
	if boardID != "" {
		query["boardId"] = boardID
	}
	return c.Request("GET", fmt.Sprintf("/price/%s/trades/latest", symbol), query, nil, nil, dryRun)
}

// GetClosePrice retrieves the close price.
func (c *Client) GetClosePrice(symbol, boardID string, dryRun bool) (int, []byte, error) {
	query := make(map[string]string)
	if boardID != "" {
		query["boardId"] = boardID
	}
	return c.Request("GET", fmt.Sprintf("/price/%s/close", symbol), query, nil, nil, dryRun)
}

File: rust/src/rest/api_market.rs
use super::client::DnseClient;
use reqwest::Method;
use std::collections::HashMap;

impl DnseClient {
    pub async fn get_security_definition(&self, symbol: &str, board_id: Option<&str>, dry_run: bool) -> Result<(u16, Vec<u8>), Box<dyn std::error::Error>> {
        let mut query = HashMap::new();
        if let Some(b) = board_id {
            query.insert("boardId", b.to_string());
        }
        self.request(Method::GET, &format!("/price/{}/secdef", symbol), Some(&query), None, None, dry_run).await
    }

    pub async fn get_ohlc(&self, bar_type: &str, mut query_params: HashMap<&str, String>, dry_run: bool) -> Result<(u16, Vec<u8>), Box<dyn std::error::Error>> {
        query_params.insert("type", bar_type.to_string());
        self.request(Method::GET, "/price/ohlc", Some(&query_params), None, None, dry_run).await
    }

    pub async fn get_trades(&self, symbol: &str, board_id: Option<&str>, from_date: Option<&str>, to_date: Option<&str>, limit: Option<usize>, order: Option<&str>, next_page_token: Option<&str>, dry_run: bool) -> Result<(u16, Vec<u8>), Box<dyn std::error::Error>> {
        let mut query = HashMap::new();
        if let Some(b) = board_id { query.insert("boardId", b.to_string()); }
        if let Some(f) = from_date { query.insert("from", f.to_string()); }
        if let Some(t) = to_date { query.insert("to", t.to_string()); }
        if let Some(l) = limit { query.insert("limit", l.to_string()); }
        if let Some(o) = order { query.insert("order", o.to_string()); }
        if let Some(n) = next_page_token { query.insert("nextPageToken", n.to_string()); }

        self.request(Method::GET, &format!("/price/{}/trades", symbol), Some(&query), None, None, dry_run).await
    }

    pub async fn get_instruments(&self, symbol: Option<&str>, market_id: Option<&str>, security_group_id: Option<&str>, index_name: Option<&str>, limit: Option<usize>, page: Option<usize>, dry_run: bool) -> Result<(u16, Vec<u8>), Box<dyn std::error::Error>> {
        let mut query = HashMap::new();
        if let Some(s) = symbol { query.insert("symbol", s.to_string()); }
        if let Some(m) = market_id { query.insert("marketId", m.to_string()); }
        if let Some(g) = security_group_id { query.insert("securityGroupId", g.to_string()); }
        if let Some(i) = index_name { query.insert("indexName", i.to_string()); }
        if let Some(l) = limit { query.insert("limit", l.to_string()); }
        if let Some(p) = page { query.insert("page", p.to_string()); }

        self.request(Method::GET, "/instruments", Some(&query), None, None, dry_run).await
    }

    pub async fn get_latest_trade(&self, symbol: &str, board_id: Option<&str>, dry_run: bool) -> Result<(u16, Vec<u8>), Box<dyn std::error::Error>> {
        let mut query = HashMap::new();
        if let Some(b) = board_id {
            query.insert("boardId", b.to_string());
        }
        self.request(Method::GET, &format!("/price/{}/trades/latest", symbol), Some(&query), None, None, dry_run).await
    }

    pub async fn get_close_price(&self, symbol: &str, board_id: Option<&str>, dry_run: bool) -> Result<(u16, Vec<u8>), Box<dyn std::error::Error>> {
        let mut query = HashMap::new();
        if let Some(b) = board_id {
            query.insert("boardId", b.to_string());
        }
        self.request(Method::GET, &format!("/price/{}/close", symbol), Some(&query), None, None, dry_run).await
    }
}
